storage/remote/write_handler_test.go

```diff
@@ -290,119 +291,132 @@ func TestRemoteWriteHandler_V1Message(t *testing.T) {
+func expectHeaderValue(t testing.TB, expected int, got string) {
+	t.Helper()
 
-	req, err := http.NewRequest("", "", bytes.NewReader(payload))
+	require.NotEmpty(t, got)
+	i, err := strconv.Atoi(got)
 	require.NoError(t, err)
-
-	req.Header.Set("Content-Type", remoteWriteContentTypeHeaders[config.RemoteWriteProtoMsgV2])
-	req.Header.Set("Content-Encoding", string(SnappyBlockCompression))
-	req.Header.Set(RemoteWriteVersionHeader, RemoteWriteVersion20HeaderValue)
-
-	appendable := &mockAppendable{}
-	handler := NewWriteHandler(log.NewNopLogger(), nil, appendable, []config.RemoteWriteProtoMsg{config.RemoteWriteProtoMsgV2})
-
-	recorder := httptest.NewRecorder()
-	handler.ServeHTTP(recorder, req)
-
-	resp := recorder.Result()
-	require.Equal(t, http.StatusNoContent, resp.StatusCode)
-
-	b := labels.NewScratchBuilder(0)
-	i := 0
-	j := 0
-	k := 0
-	for _, ts := range writeV2RequestFixture.Timeseries {
-		ls := ts.ToLabels(&b, writeV2RequestFixture.Symbols)
-
-		for _, s := range ts.Samples {
-			requireEqual(t, mockSample{ls, s.Timestamp, s.Value}, appendable.samples[i])
-
-			switch i {
-			case 0:
-				requireEqual(t, mockMetadata{ls, writeV2RequestSeries1Metadata}, appendable.metadata[i])
-			case 1:
-				requireEqual(t, mockMetadata{ls, writeV2RequestSeries2Metadata}, appendable.metadata[i])
-			default:
-				t.Fatal("more series/samples then expected")
-			}
-			i++
-		}
-		for _, e := range ts.Exemplars {
-			exemplarLabels := e.ToExemplar(&b, writeV2RequestFixture.Symbols).Labels
-			requireEqual(t, mockExemplar{ls, exemplarLabels, e.Timestamp, e.Value}, appendable.exemplars[j])
-			j++
-		}
-		for _, hp := range ts.Histograms {
-			if hp.IsFloatHistogram() {
-				fh := hp.ToFloatHistogram()
-				requireEqual(t, mockHistogram{ls, hp.Timestamp, nil, fh}, appendable.histograms[k])
-			} else {
-				h := hp.ToIntHistogram()
-				requireEqual(t, mockHistogram{ls, hp.Timestamp, h, nil}, appendable.histograms[k])
-			}
-			k++
-		}
-	}
+	require.Equal(t, expected, i)
 }
 
-func TestOutOfOrderSample_V1Message(t *testing.T) {
+func TestRemoteWriteHandler_V2Message(t *testing.T) {
+	// V2 supports partial writes for non-retriable errors, so test them.
 	for _, tc := range []struct {
-		Name      string
-		Timestamp int64
+		desc             string
+		input            []writev2.TimeSeries
+		expectedCode     int
+		expectedRespBody string
+
+		commitErr          error
+		appendSampleErr    error
+		appendHistogramErr error
+		appendExemplarErr  error
+		updateMetadataErr  error
 	}{
 		{
-			Name:      "historic",
-			Timestamp: 0,
+			desc:         "All timeseries accepted",
+			input:        writeV2RequestFixture.Timeseries,
+			expectedCode: http.StatusNoContent,
 		},
 		{
-			Name:      "future",
-			Timestamp: math.MaxInt64,
+			desc: "Partial write; first series with invalid labels (no metric name)",
+			input: append(
+				[]writev2.TimeSeries{{LabelsRefs: []uint32{2, 2}, Samples: []writev2.Sample{{Value: 1, Timestamp: 1}}}},
```

Comment:
without knowing what the contents of the symbol table are it's hard to know what `2, 2` are referencing, same for the rest of the tests

Reply:
Fair. Added comments. 👍🏽